### PR TITLE
Initial draft for device properties structure

### PR DESCRIPTION
**Pull request by Tomatosoup97**
Early draft of how the `devprops` (device properties) could look like. Any critique or improvement suggestion is welcome.

### CODE REVIEW

File: include/device.h [hunk 1]
```diff
@@ -41,15 +74,20 @@ struct device {
   /* Device information and state. */
   device_bus_t bus;
   driver_t *driver;
+  // TODO: most likeley we want to get rid of `instance` field
   void *instance; /* used by bus driver to store data in children */
   void *state;    /* memory requested by driver for its state*/
+  device_attr_t attributes[DEV_ATTR_MAX]; /* equivalent of FreeBSD's `ivars` */
```
Comment (mzr) [inline include/device.h]: Now that comment is not exactly true and I think it's more confusing than helpful.

File: include/device.h [hunk 2]
```diff
@@ -30,6 +30,39 @@ struct driver {
...
+  DA_FDT_PATH,
+  /* resource types */
+  DA_IOPORT,
+  DA_MEMORY,
+  DA_IRQ,
+  /* PCI bus identifiers */
+  DA_PCI_CLASSCODE,
+  DA_PCI_DEVICEID,
+  DA_PCI_VENDORID,
+} __packed da_tag_t;
```
Comment (mzr) [inline include/device.h]: I'm not sure about this one, but is it possible that we may encounter unaligned access when using `__packed`?

Comment (cahirwpz) [reply include/device.h]: Packed unions will take the least space that can accommodate all members => in this case `char` will do. Normally it'd consume space for `int`.

File: tests/device.c [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+#include <condvar.h>
+#include <ringbuf.h>
+#include <pci.h>
+#include <dev/isareg.h>
+#include <dev/ns16550reg.h>
+#include <interrupt.h>
+#include <sysinit.h>
+
+#if 0
+extern device_t *gt_pci;
```
Comment (mzr) [inline tests/device.c]: Probably we should get reference to this device in some other, dynamic way.

**Comment by cahirwpz** [comment]
This PR got heavily outdated. There's too little value in it to be worth spending time to update it to current code base. Hence, I decided to close it.

**Pull request closed without merging**